In [1]:
#importing important packages
import pandas as pd
import numpy as np

#reading the dataset
df = pd.read_csv(r'H:\Sandeep\DATA SCIENCE\MY Files\Ensemble classifier\train.csv')

#filling missing values
# df['Gender'].fillna('Male', inplace=True)

In [2]:
df

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

In [3]:
df.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [4]:
#filling missing values
df['Gender'].fillna('Male', inplace=True)
df['Married'].fillna('Yes', inplace=True)
df['Dependents'].fillna(0, inplace=True)
df['Self_Employed'].fillna('No', inplace=True)
df['LoanAmount'].fillna(128.0, inplace=True)
df['Loan_Amount_Term'].fillna(342.0, inplace=True)
df['Credit_History'].fillna(1.0 , inplace=True)


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             614 non-null    object 
 2   Married            614 non-null    object 
 3   Dependents         614 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      614 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         614 non-null    float64
 9   Loan_Amount_Term   614 non-null    float64
 10  Credit_History     614 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [6]:
df['Dependents'].replace({'3+':3},inplace=True)


In [7]:
df['Dependents']=df['Dependents'].astype(int)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             614 non-null    object 
 2   Married            614 non-null    object 
 3   Dependents         614 non-null    int32  
 4   Education          614 non-null    object 
 5   Self_Employed      614 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         614 non-null    float64
 9   Loan_Amount_Term   614 non-null    float64
 10  Credit_History     614 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int32(1), int64(1), object(7)
memory usage: 60.1+ KB


In [9]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(df, test_size=0.3, random_state=0)
x_train=train.drop(['Loan_ID','Loan_Status'],axis=1)
y_train=train['Loan_Status']

x_test=test.drop(['Loan_ID','Loan_Status'],axis=1)
y_test=test['Loan_Status']

x_test.shape,y_test.shape


((185, 11), (185,))

In [10]:
x_train=pd.get_dummies(x_train)
x_test=pd.get_dummies(x_test)
x_test.shape,y_test.shape

((185, 17), (185,))

# Bagging meta-estimator

In [11]:
from sklearn.ensemble import BaggingClassifier
from sklearn import tree
model = BaggingClassifier(tree.DecisionTreeClassifier(random_state=1))
model.fit(x_train, y_train)
model.score(x_test,y_test)


0.7783783783783784

# Random Forest

In [12]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score


In [13]:
model = RandomForestClassifier()
model.fit(x_train, y_train)

RandomForestClassifier()

In [14]:
print('Number of Trees used : ', model.n_estimators)

Number of Trees used :  100


In [15]:
predict_train = model.predict(x_train)
print('\nTarget on train data',predict_train) 


Target on train data ['N' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'N' 'Y' 'N' 'N' 'Y'
 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'N' 'Y' 'N' 'Y' 'Y' 'Y'
 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'N' 'N' 'N' 'Y' 'Y' 'Y' 'N' 'Y' 'N'
 'N' 'Y' 'N' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'N' 'N' 'Y' 'Y' 'N' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N'
 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'N' 'Y' 'N' 'N' 'N' 'Y' 'N' 'Y' 'Y' 'Y'
 'Y' 'Y' 'N' 'N' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'N'
 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'N' 'Y' 'Y' 'Y'
 'N' 'Y' 'Y' 'Y' 'N' 'Y' 'N' 'Y' 'N' 'Y' 'Y' 'N' 'N' 'N' 'Y' 'N' 'Y' 'Y'
 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'N' 'Y' 'N' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'N' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'N' 'Y'
 'N' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'N' 'Y' 'Y'
 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 

In [16]:
accuracy_train = accuracy_score(y_train,predict_train)
print('\naccuracy_score on train dataset : ', accuracy_train)


accuracy_score on train dataset :  1.0


In [17]:
predict_test = model.predict(x_test)
print('\nTarget on test data',predict_test)


Target on test data ['Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'N' 'Y' 'Y' 'Y' 'N'
 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'N' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'N'
 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y'
 'Y' 'N' 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'N' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y']


In [18]:
accuracy_test = accuracy_score(y_test,predict_test)
print('\naccuracy_score on test dataset : ', accuracy_test)


accuracy_score on test dataset :  0.8


# AdaBoost

In [19]:
from sklearn.ensemble import AdaBoostClassifier
model = AdaBoostClassifier(random_state=1)
model.fit(x_train, y_train)
model.score(x_test,y_test)


0.8162162162162162

# Gradient Boosting 

In [20]:
from sklearn.ensemble import GradientBoostingClassifier
model= GradientBoostingClassifier(learning_rate=0.01,random_state=1)
model.fit(x_train, y_train)
model.score(x_test,y_test)


0.8162162162162162

# XGBoost

In [35]:
y_train.unique()

array(['N', 'Y'], dtype=object)

In [36]:
y_train.value_counts()

Loan_Status
Y    288
N    141
Name: count, dtype: int64

In [40]:
y_train_=y_train.apply(lambda x:1  if x=='Y' else 0)
y_test_=y_test.apply(lambda x:1  if x=='Y' else 0)


In [41]:
y_train_.unique()

array([0, 1], dtype=int64)

In [46]:
y_train_.value_counts()

Loan_Status
1    288
0    141
Name: count, dtype: int64

In [42]:
y_test_.unique()

array([1, 0], dtype=int64)

In [47]:
y_test_.value_counts()

Loan_Status
1    134
0     51
Name: count, dtype: int64

In [45]:
import xgboost as xgb
model=xgb.XGBClassifier(random_state=1,learning_rate=0.01)
model.fit(x_train, y_train_)
model.score(x_test,y_test_)

0.8216216216216217

# Project Summary: Ensemble Algorithms for Loan Status Prediction
In this project, I implemented and evaluated several ensemble algorithms to predict loan status (Yes or No) using a binary classification approach. The ensemble methods used included XGBoost, Gradient Boosting, AdaBoost, Random Forest, and Bagging. The objective was to determine the most effective algorithm in terms of accuracy for this specific predictive task.

Algorithms and Results:
XGBoost (Extreme Gradient Boosting):

Accuracy: 0.821
XGBoost provided the highest accuracy among all the models, showcasing its powerful ability to handle complex patterns in the data and effectively reduce overfitting.
Gradient Boosting:

Accuracy: 0.816
Gradient Boosting also performed exceptionally well, closely following XGBoost. It iteratively builds models and corrects errors of the previous models, which enhances its predictive capability.
AdaBoost (Adaptive Boosting):

Accuracy: 0.816
AdaBoost demonstrated the same accuracy as Gradient Boosting. It works by combining multiple weak classifiers to form a strong classifier, focusing more on the misclassified instances in each iteration.
Random Forest:

Accuracy: 0.80
Random Forest, with its ensemble of decision trees, provided a robust and reliable accuracy. It benefits from the randomness introduced in both feature selection and data sampling.
Bagging (Bootstrap Aggregating):

Accuracy: 0.77
Bagging, while still effective, showed the lowest accuracy among the evaluated models. It reduces variance by training multiple models on different random subsets of the training data and then averaging the predictions.

# Key Steps in the Project:
Data Preprocessing: Encoding categorical columns, standardization, and handling missing values.
Feature Engineering: Selecting and transforming relevant features to improve model performance.
Model Training and Evaluation: Implementing and training each ensemble algorithm, followed by evaluation using accuracy metrics.
Visualization: Creating visual representations of model performance and feature importance to gain insights.

# Conclusion:
This project demonstrated the effectiveness of various ensemble algorithms in predicting loan status. XGBoost emerged as the most accurate model, followed by Gradient Boosting and AdaBoost, both of which also showed high accuracy. Random Forest and Bagging, while slightly less accurate, still provided valuable insights and robust predictions. Ensemble methods proved to be powerful tools for improving prediction accuracy and model robustness in binary classification tasks.